File: 6216.py
class Solution:
    def minCost(self, nums: List[int], cost: List[int]) -> int:
        def total_cost(avg):
            return sum(abs(n - avg) * c for (n, c) in zip(nums, cost))
        left, right = MN, MX = min(nums), max(nums)
        while left < right:
            mid = (left + right) // 2
            lres = float("inf") if mid == MN else total_cost(mid - 1)
            mres = total_cost(mid)
            rres = float("inf") if mid == MX else total_cost(mid + 1)
            if lres < mres:
                right = mid - 1
            elif rres < mres:
                left = mid + 1
            else:
                left = right = mid
        return total_cost(left)

File: 6216.c
long long totalCost( int avg, int* nums, int* cost, int size) {
    long long res = 0;
    for (int i = 0; i < size; i++) {
        res += abs(nums[i] - avg) * (long long)cost[i];
    }
    return res;
}

long long minCost(int* nums, int numsSize, int* cost, int costSize) {
    const int MN = 0, MX = 1000000;
    int left = MN, right = MX;
    while (left < right) {
        int mid = (left + right) / 2;
        long long mres = totalCost(mid, nums, cost, numsSize);
        if (mres > totalCost(mid - 1, nums, cost, numsSize)) right = mid - 1;
        else if (mres > totalCost(mid + 1, nums, cost, numsSize)) left = mid + 1;
        else left = right = mid;
    }
    return totalCost(left, nums, cost, numsSize);
}

